File: Data Structure/Linked List/Medium/LET_no725_Split Linked List in Parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int len = 0;
        for (ListNode* cur = head; cur; cur = cur->next)
            len++;
        int n = len / k, rest = len % k;
        ListNode* last = nullptr;
        vector<ListNode*> res(k, nullptr);
        
        for (int i = 0; head && i < k; i++, rest--) {
            res[i] = head;
            for (int j = 0; j < n + (rest > 0); j++) {
                last = head;
                head = head->next;
            }
            last->next = nullptr;
        }
        
        return res;
    }
};

File: Data Structure/Linked List/Medium/LET_no725_Split Linked List in Parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        cur = head
        n = 0
        while cur:
            n += 1
            cur = cur.next
        divide, rest = n // k, n % k
        res = []
        last = None
        for _ in range(k):
            res.append(head)
            times = divide + (rest > 0)
            while times and head:
                last = head
                head = head.next
                times -= 1
            rest -= 1
            if last: last.next = None
        return res